### DISCUSSION TITLE
Tree shaking eliminates only part of package's code

### DISCUSSION THREAD

**Discussion by yishayweb** [webpack/webpack#17073] [Q&A]
Hi

I have the following scenario: I created a package of react components (which uses mui and the styled utility exported from @mui/material/styles). I bundle this package with rollup. Then, I'm using the package in an app which I bundle with webpack.
The package.json of the package includes **sideEffects: false**

In the app, I use only one of the two exported components I have in the package. The component that I **don't** use (called TextWithIcon) is comprised of a .tsx file that includes the react component, and a .styles.tsx file that includes the styled object. The .tsx file imports the .styles.tsx file.

When I build the app in production mode, I notice that the bundle **excludes** (as expected) the code of the actual component (the code in the .tsx file) but **includes** the code in the styled object (the code in the .styles.tsx file) although this code is never used.

So I wonder

1. Why does the .styles.tsx code remain in the webpack bundle if I don't use it?
2. I tried to avoid using rollup, and just transpiled the code of the package with tsc, and then the unused code was indeed eliminated. So why with un-bundled code webpack knows to eliminate the unused code?

App code: [here](https://github.com/yishayweb/app-example)
Package bundled with rollup code: [here](https://github.com/yishayweb/package-example)

Would really appreciate if someone can clarify this

Thanks !!

**Comment by alexander-akait** [comment]
Try to enable this for CSS loaders https://webpack.js.org/configuration/module/#rulesideeffects

**Reply by yishayweb** [reply]
Thanks for the quick response @alexander-akait
I tried to use the sideEffects option in optimization mentioned [here](https://github.com/webpack/webpack.js.org/blob/main/src/content/configuration/optimization.mdx#optimizationsideeffects) with 'true' value as explained, and the one you suggested mentioned [here](https://github.com/webpack/webpack.js.org/blob/main/src/content/configuration/module.mdx#rulesideeffects) with true/false values because it wasn't clear what the value should be, but non of the options worked
Also, the styled code is for styles but it is still in tsx file (because it's css in js) so I'm not sure it is relevant to css loaders. I tried to enable the flag anyway in all kinds of loaders

Do you have any idea why it's not working or any other suggestions?

**Reply by alexander-akait** [reply]
Need to look at generated code (I don't loomat your examples currently), you can try to add `/*@__PURE__*/` (more information - https://github.com/terser/terser#annotations), but maybe we just don't handle this scenario

**Reply by yishayweb** [reply]
I tried to add the `/*@__PURE__*/` annotation before the unused styled function and indeed it eliminates the code properly.
I think I figured out this behaviour.
I saw in couple of references that the **sideEffects: false** flag doesn't work on single bundle packages because it eliminates whole files and not just part of them. So it looks like it doesn't influence in this case. The only tree shake I get is thanks to the **usedExports: true** flag which is used by terser.
In the example package I created, the code in the .tsx file (the actual component) is being eliminated by terser, but the styled function the component uses isn't being eliminated, and this I guess is caused by some terser rules.

@alexander-akait do you know if terser somehow identifies the styled function of material-ui and the style rules inside it as kind of css and hence (without an explicit `/*@__PURE__*/` annotation) it doesn't eliminate it?

**Reply by alexander-akait** [reply]
By default any function is not pure, because terser doesn't know about your function nothing, do you have babel (or something else) plugin, because it should add `/*@__PURE__*/` (or you can implement this plugin :smile: )

**Reply by yishayweb** [reply]
But if terser doesn't know then how does it remove the code of the component (the TextWithIcon, see below)? The code there also doesn't have `/*@__PURE__*/` annotation?

This is the code in the bundle of the package. The TextIconStyle is not removed but TextWithIcon is removed. Both of them are unused in the app. Then why the difference?

```
const TextIconStyle = styled$1('div', {
    shouldForwardProp: (prop) => !['hoverFlag', 'iconLocation'].includes(prop),
})(({ hoverFlag, iconLocation }) => ({
    padding: '5px 15px 5px',
    lineHeight: '1.5rem',
    '& .this-is-some-class': {
        alignSelf: 'center',
        color: 'green',
    },
}));

const TextWithIcon = ({ iconLocation = 'left', text, onClick, }) => {
    return (jsxs(TextIconStyle, Object.assign({ hoverFlag: !!onClick, iconLocation: iconLocation }, { children: ["Hi the icon with text:", jsx("div", Object.assign({ className: "this-is-some-class" }, { children: text }), void 0)] }), void 0));
};
```

@alexander-akait I'm sorry for insisting on this, I'm just curious about it :)

**Reply by alexander-akait** [reply]
> But if terser doesn't know then how does it remove the code of the component (the TextWithIcon, see below)? The code there also doesn't have /*@__PURE__*/ annotation?

Terser is just a minifier, it doesn't know nothing about `sideEffects: false`, because it is on the webpack level

there is the `pure_funcs` option in terser

> pure_funcs (default: null) -- You can pass an array of names and Terser will assume that those functions do not produce side effects. DANGER: will not check if the name is redefined in scope. An example case here, for instance var q = Math.floor(a/b). If variable q is not used elsewhere, Terser will drop it, but will still keep the Math.floor(a/b), not knowing what it does. You can pass pure_funcs: [ 'Math.floor' ] to let it know that this function won't produce any side effect, in which case the whole statement would get discarded. The current implementation adds some overhead (compression will be slower).


Anyway give me a time, I will look deeply

**Reply by alexander-akait** [reply]
hm, I tested your example in production mode and can't find `TextWithIcon` in chunks, maybe I missing something, how you find it?

**Reply by yishayweb** [reply]
@alexander-akait I made some experiments with terser and I think I understand now the behavior I get.
From what I've seen, terser knows to analyze js expressions until certain level of complexity. If it succeeds to analyze an expression and deduce that for sure it has no other code imports it and that it has no side effects, then it removes it (and the variable that holds it). But when the expression is too complex and terser can't deduce that there are no side effects, it doesn't remove it.

I may be wrong, but that is I think what happend with the TextIconStyle variable I have in the code, which holds a rather complex expression.

If you intend to look into it sometime, I would love to know your conclusions.
Anyway, thanks for responding me and for the tips you gave me along the way. I learned a lot and I really appriciate it :)

**Reply by yishayweb** [reply]
I saw only now the comment about you testing the code.
You are right, `TextWithIcon` is not in the chunks. It is removed. But `TextIconStyle` is remained although it is unused

I would love to hear your thoughts on the previous comment I made about expressions complexity. Does it sound right to you?

**Reply by yishayweb** [reply]
I read more into details [this part of tree shaking docs](https://webpack.js.org/guides/tree-shaking/#clarifying-tree-shaking-and-sideeffects) and I think it's validating what I wrote [in the previous comment](https://github.com/webpack/webpack/discussions/17073#discussioncomment-5758977)
In the docs it says that terser has difficulties determining if code like this `withAppProvider()(Button)` (react HOC) has side effects or not.
From the docs:
`Terser actually tries to figure it out, but it doesn't know for sure in many cases`

This HOC case from the docs is similar to the `TextIconStyle` I asked about because it holds an expression that is something like this `styled$1(someArg)(someOtherArg)`. It has the same structure like the HOC case in the docs.

Thanks again @alexander-akait for helping me out with this :)
Should I mark this discussion as 'Answered'?

**Reply by alexander-akait** [reply]
Yeah, I am afraid we can't solve it without `/*@__PURE__*/` or `pure_funcs`, I think you can create a small babel to add pure comments
